src/plugins/spatial_grid.rs

```diff
@@ -0,0 +1,482 @@
+    {
+        // Remove entity from old position if it exists
+        self.remove(entity);
+
+        // Insert at new position
+        self.position_to_entities
+            .entry(position)
+            .or_default()
+            .insert(entity);
+        self.entity_to_position.insert(entity, position);
+    }
+
+    /// Remove an entity from the spatial index.
+    ///
+    /// Returns the position where the entity was located, if it was found.
+    pub(crate) fn remove(&mut self, entity: Entity) -> Option<GridPosition<T>>
+    {
+        if let Some(position) = self.entity_to_position.remove(&entity)
+            && let Some(entities) = self.position_to_entities.get_mut(&position)
+        {
+            entities.remove(&entity);
+            if entities.is_empty()
+            {
+                self.position_to_entities.remove(&position);
+            }
+            Some(position)
+        }
+        else
+        {
+            None
+        }
+    }
+
+    /// Get all entities at a specific position.
+    pub fn entities_at(&self, position: &GridPosition<T>) -> impl Iterator<Item = Entity> + '_
+    {
+        self.position_to_entities
+            .get(position)
+            .into_iter()
+            .flat_map(|set| set.iter().copied())
+    }
+
+    /// Get the position of an entity.
+    #[must_use]
+    pub fn position_of(&self, entity: Entity) -> Option<GridPosition<T>>
+    {
+        self.entity_to_position.get(&entity).copied()
+    }
+
+    /// Get all entities in the neighborhood of a position (Moore neighborhood).
+    pub fn neighbors_of(&self, position: &GridPosition<T>) -> impl Iterator<Item = Entity>
+    {
+        position
+            .0
+            .neighbors()
+            .filter(|neighbor_pos| {
+                self.bounds
+                    .is_none_or(|bounds| neighbor_pos.in_bounds(&bounds))
+            })
+            .map(|p| GridPosition(p))
+            .flat_map(|neighbor_pos| {
+                self.position_to_entities
+                    .get(&neighbor_pos)
+                    .into_iter()
+                    .flat_map(|set| set.iter().copied())
+            })
+    }
+
+    /// Get all entities in the orthogonal neighborhood of a position (Von Neumann neighborhood).
+    pub fn orthogonal_neighbors_of(
+        &self,
+        position: &GridPosition<T>,
+    ) -> impl Iterator<Item = Entity>
+    {
+        position
+            .0
+            .neighbors_orthogonal()
+            .filter(|neighbor_pos| {
+                self.bounds
+                    .is_none_or(|bounds| neighbor_pos.in_bounds(&bounds))
+            })
+            .map(|p| GridPosition(p))
+            .flat_map(|neighbor_pos| {
+                self.position_to_entities
+                    .get(&neighbor_pos)
+                    .into_iter()
+                    .flat_map(|set| set.iter().copied())
+            })
+    }
+
+    /// Clear all entities from the spatial index.
+    pub fn clear(&mut self)
```

Comment:
Should be private.
Same for `insert` and `remove` actually, I think `pub(crate)` is not necessary.